File: data-jpa/src/main/java/top/bettercode/summer/data/jpa/query/mybatis/TupleResultSetMetaData.java
package top.bettercode.summer.data.jpa.query.mybatis;

import java.sql.ResultSetMetaData;
import java.sql.SQLException;
import java.sql.SQLFeatureNotSupportedException;
import java.util.ArrayList;
import java.util.List;
import javax.persistence.Tuple;
import javax.persistence.TupleElement;
import org.apache.ibatis.type.JdbcType;
import org.hibernate.type.descriptor.sql.JdbcTypeJavaClassMappings;

/**
 * @author Peter Wu
 */
public class TupleResultSetMetaData implements ResultSetMetaData {

  private final List<String> columnNames = new ArrayList<>();
  private final List<String> classNames = new ArrayList<>();
  private final List<JdbcType> jdbcTypes = new ArrayList<>();

  public TupleResultSetMetaData(List<Tuple> tuples) {
    for (Tuple tuple : tuples) {
      List<TupleElement<?>> elements = tuple.getElements();
      for (int i = 0; i < elements.size(); i++) {
        TupleElement<?> element = elements.get(i);
        String alias = element.getAlias();
        columnNames.add(i, alias);
        Class<?> javaType = element.getJavaType();
        if (!javaType.equals(Object.class)) {
          classNames.add(i, javaType.getName());
          int type = JdbcTypeJavaClassMappings.INSTANCE.determineJdbcTypeCodeForJavaClass(javaType);
          JdbcType jdbcType = JdbcType.forCode(type);
          jdbcTypes.add(i, jdbcType);
        } else if (classNames.size() <= i) {
          classNames.add(i, javaType.getName());
          jdbcTypes.add(i, JdbcType.VARCHAR);
        }
      }
    }
  }

  public int findColumn(String columnName) {
    return columnNames.indexOf(columnName);
  }

  @Override
  public int getColumnCount() throws SQLException {
    return columnNames.size();
  }

  @Override
  public boolean isAutoIncrement(int i) throws SQLException {
    throw new SQLFeatureNotSupportedException();
  }

  @Override
  public boolean isCaseSensitive(int i) throws SQLException {
    throw new SQLFeatureNotSupportedException();
  }

  @Override
  public boolean isSearchable(int i) throws SQLException {
    throw new SQLFeatureNotSupportedException();
  }

  @Override
  public boolean isCurrency(int i) throws SQLException {
    throw new SQLFeatureNotSupportedException();
  }

  @Override
  public int isNullable(int i) throws SQLException {
    throw new SQLFeatureNotSupportedException();
  }

  @Override
  public boolean isSigned(int i) throws SQLException {
    throw new SQLFeatureNotSupportedException();
  }

  @Override
  public int getColumnDisplaySize(int i) throws SQLException {
    throw new SQLFeatureNotSupportedException();
  }

  @Override
  public String getColumnLabel(int i) throws SQLException {
    return columnNames.get(i - 1);
  }

  @Override
  public String getColumnName(int i) throws SQLException {
    return columnNames.get(i - 1);
  }

  @Override
  public String getSchemaName(int i) throws SQLException {
    throw new SQLFeatureNotSupportedException();
  }

  @Override
  public int getPrecision(int i) throws SQLException {
    throw new SQLFeatureNotSupportedException();
  }

  @Override
  public int getScale(int i) throws SQLException {
    throw new SQLFeatureNotSupportedException();
  }

  @Override
  public String getTableName(int i) throws SQLException {
    throw new SQLFeatureNotSupportedException();
  }

  @Override
  public String getCatalogName(int i) throws SQLException {
    throw new SQLFeatureNotSupportedException();
  }

  @Override
  public int getColumnType(int i) throws SQLException {
    return jdbcTypes.get(i - 1).TYPE_CODE;
  }

  @Override
  public String getColumnTypeName(int i) throws SQLException {
    throw new SQLFeatureNotSupportedException();
  }

  @Override
  public boolean isReadOnly(int i) throws SQLException {
    throw new SQLFeatureNotSupportedException();
  }

  @Override
  public boolean isWritable(int i) throws SQLException {
    throw new SQLFeatureNotSupportedException();
  }

  @Override
  public boolean isDefinitelyWritable(int i) throws SQLException {
    throw new SQLFeatureNotSupportedException();
  }

  @Override
  public String getColumnClassName(int i) throws SQLException {
    return classNames.get(i - 1);
  }

  @Override
  public <T> T unwrap(Class<T> aClass) throws SQLException {
    throw new SQLFeatureNotSupportedException();
  }

  @Override
  public boolean isWrapperFor(Class<?> aClass) throws SQLException {
    throw new SQLFeatureNotSupportedException();
  }
}

File: data-jpa/src/main/kotlin/top/bettercode/summer/data/jpa/query/mybatis/TupleResultSetMetaData.kt
package top.bettercode.summer.data.jpa.query.mybatis

import org.apache.ibatis.type.JdbcType
import java.sql.ResultSetMetaData
import java.sql.SQLFeatureNotSupportedException
import javax.persistence.Tuple

/**
 * @author Peter Wu
 */
class TupleResultSetMetaData(tuples: List<Tuple>) : ResultSetMetaData {
    private val columnNames: MutableList<String> = ArrayList()
    private val classNames: MutableList<String> = ArrayList()
    private val jdbcTypes: MutableList<JdbcType> = ArrayList()

    init {
        val size = tuples.size
        if (size > 0) {
            val firstTuple = tuples[0]
            val firstElements = firstTuple.elements
            val indexes = mutableListOf<Int>()
            firstElements.indices.forEach { i ->
                val element = firstElements[i] as NativeTupleElementImpl
                columnNames.add(element.alias)
                classNames.add(element.javaType.name)
                val jdbcType = element.getJdbcType()
                jdbcTypes.add(jdbcType)
                if (jdbcType == JdbcType.NULL) {
                    indexes.add(i)
                }
            }

            if (indexes.isNotEmpty()) {
                for (t in 1 until size) {
                    val tuple = tuples[t]
                    val elements = tuple.elements
                    val iterator = indexes.iterator()
                    iterator.forEach { index ->
                        val element = elements[index] as NativeTupleElementImpl
                        val jdbcType = element.getJdbcType()
                        if (jdbcType != JdbcType.NULL) {
                            classNames[index] = element.javaType.name
                            jdbcTypes[index] = jdbcType
                            iterator.remove()
                        }
                    }
                    if (indexes.isEmpty()) {
                        break
                    }
                }
            }
        }
    }

    fun findColumn(columnName: String): Int {
        return columnNames.indexOf(columnName)
    }

    override fun getColumnCount(): Int {
        return columnNames.size
    }

    override fun isAutoIncrement(i: Int): Boolean {
        throw SQLFeatureNotSupportedException()
    }

    override fun isCaseSensitive(i: Int): Boolean {
        throw SQLFeatureNotSupportedException()
    }

    override fun isSearchable(i: Int): Boolean {
        throw SQLFeatureNotSupportedException()
    }

    override fun isCurrency(i: Int): Boolean {
        throw SQLFeatureNotSupportedException()
    }

    override fun isNullable(i: Int): Int {
        throw SQLFeatureNotSupportedException()
    }

    override fun isSigned(i: Int): Boolean {
        throw SQLFeatureNotSupportedException()
    }

    override fun getColumnDisplaySize(i: Int): Int {
        throw SQLFeatureNotSupportedException()
    }

    override fun getColumnLabel(i: Int): String {
        return columnNames[i - 1]
    }

    override fun getColumnName(i: Int): String {
        return columnNames[i - 1]
    }

    override fun getSchemaName(i: Int): String {
        throw SQLFeatureNotSupportedException()
    }

    override fun getPrecision(i: Int): Int {
        throw SQLFeatureNotSupportedException()
    }

    override fun getScale(i: Int): Int {
        throw SQLFeatureNotSupportedException()
    }

    override fun getTableName(i: Int): String {
        throw SQLFeatureNotSupportedException()
    }

    override fun getCatalogName(i: Int): String {
        throw SQLFeatureNotSupportedException()
    }

    override fun getColumnType(i: Int): Int {
        return jdbcTypes[i - 1].TYPE_CODE
    }

    override fun getColumnTypeName(i: Int): String {
        throw SQLFeatureNotSupportedException()
    }

    override fun isReadOnly(i: Int): Boolean {
        return true
    }

    override fun isWritable(i: Int): Boolean {
        return false
    }

    override fun isDefinitelyWritable(i: Int): Boolean {
        return false
    }

    override fun getColumnClassName(i: Int): String {
        return classNames[i - 1]
    }

    override fun <T> unwrap(iface: Class<T>): T {
        return iface.cast(this)
    }

    override fun isWrapperFor(aClass: Class<*>): Boolean {
        return aClass.isInstance(this)
    }
}
